polynote-kernel/src/main/scala/polynote/kernel/util/NotebookContext.scala

```diff
@@ -0,0 +1,155 @@
+              case ctx => ctx.id
+            }.toSet
+
+            cellContext.setPrev(at)
+
+            tail.headOption.foreach(_.setPrev(cellContext))
+
+            tail.filter(_.previous.exists(ctx => cellsBefore(ctx.id))).foreach {
+              ctx => ctx.setPrev(cellContext)
+            }
+
+            if (at eq _last) {
+              _last = cellContext
+            }
+
+            cellContext
+        }
+      }
+  }
+
+  def insertLast(cellContext: CellContext): CellContext = synchronized {
+    if (_last != null) {
+      cellContext.setPrev(_last)
+      _last = cellContext
+      cellContext
+    } else {
+      assert(_first == null, "Broken links")
+      _first = cellContext
+      _last = cellContext
+      cellContext
+    }
+  }
+
+  def insertFirst(cellContext: CellContext): CellContext = synchronized {
+    if (_first != null) {
+      _first.setPrev(cellContext)
+      _first = cellContext
+      cellContext
+    } else {
+      assert(_last == null, "Broken links")
+      _first = cellContext
+      _last = cellContext
+      cellContext
+    }
+  }
+
+  def remove(id: CellID): Unit = synchronized {
+    var last = this._last
+
+    if (last != null && last.id == id) {
+      if (last.previous.isDefined) {
+        this._last = last.previous.get
+      } else {
+        assert(this._first == last, "Broken links")
+        this._last = null
+        this._first = null
+      }
+    } else {
+      if (last != null) {
+        var nextToLast: CellContext = last
+        last = last.previous.orNull
+
+        while (last != null) {
+          if (last.id == id) {
+            nextToLast.setPrev(last.previous)
+            return
+          }
+          nextToLast = last
+          last = last.previous.orNull
+        }
+      }
+      // not found
+      throw new NoSuchElementException(s"No cell context with id $id to remove")
+    }
+  }
+
+  def get(id: CellID): Option[CellContext] = find(_.id == id)
+  def getIO(id: CellID): IO[CellContext] = IO(find(_.id == id)).flatMap {
+    case None => IO.raiseError(new NoSuchElementException(s"Cell $id"))
+    case Some(ctx) => IO.pure(ctx)
+  }
+
+  def find(fn: CellContext => Boolean): Option[CellContext] = {
+    @tailrec def impl(ctxOpt: Option[CellContext]): Option[CellContext] = ctxOpt match {
+      case some@Some(ctx) if fn(ctx) => some
+      case Some(ctx) => impl(ctx.previous)
+      case None => None
+    }
+    impl(Option(_last))
+  }
+
+  def first: Option[CellContext] = Option(_first)
+  def last: Option[CellContext] = Option(_last)
+
+  def allResultValues: List[ResultValue] = Option(_last).toList.flatMap(_.visibleValues)
+
+  /**
+    * If the context is already present, replace it with the given context and return old context. Otherwise, return None.
+    */
+  def tryUpdate(context: CellContext): Option[CellContext] = synchronized {
+    find(_.previous.exists(_.id == context.id)) match {
+      case Some(successor) =>
```

Comment:
So since we preload the NotebookContext in `PolyKernel.init()` we should be in this case most of the time - basically aside from a fresh notebook?